File: backend/go-chi-hexagonal/internal/application/service/article_service.go
package service

import (
	"context"
	"time"

	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hexagonal/internal/application/port"
	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hexagonal/internal/domain"
)

type articleService struct {
	articleRepo port.ArticleRepository
	userRepo    port.UserRepository
}

func NewArticleService(articleRepo port.ArticleRepository, userRepo port.UserRepository) port.ArticleService {
	return &articleService{
		articleRepo: articleRepo,
		userRepo:    userRepo,
	}
}

func (s *articleService) CreateArticle(ctx context.Context, cmd port.CreateArticleCommand) (*domain.Article, error) {
	author, err := s.userRepo.FindByID(ctx, cmd.AuthorID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if author == nil {
		return nil, domain.NewResourceNotFound("User", "id", cmd.AuthorID)
	}

	// Check if article with same title already exists
	existingArticle, err := s.articleRepo.GetByTitle(ctx, cmd.Title, nil)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if existingArticle != nil {
		return nil, domain.NewAlreadyExistsError("Article with this title already exists")
	}

	slug := domain.Slugify(cmd.Title)

	// Check if article with same slug already exists
	existingArticle, err = s.articleRepo.GetBySlug(ctx, slug, nil)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if existingArticle != nil {
		return nil, domain.NewAlreadyExistsError("Article with this slug already exists")
	}

	article := &domain.Article{
		Slug:           slug,
		Title:          cmd.Title,
		Description:    cmd.Description,
		Body:           cmd.Body,
		TagList:        cmd.TagList,
		CreatedAt:      time.Now(),
		UpdatedAt:      time.Now(),
		Favorited:      false,
		FavoritesCount: 0,
		Author: domain.Profile{
			Username:  author.Username,
			Bio:       author.Bio,
			Image:     author.Image,
			Following: false,
		},
	}

	if err := s.articleRepo.Create(ctx, article, author.ID); err != nil {
		return nil, domain.NewInternalError(err.Error())
	}

	return article, nil
}

func (s *articleService) GetArticle(ctx context.Context, query port.GetArticleQuery) (*domain.Article, error) {
	article, err := s.articleRepo.GetBySlug(ctx, query.Slug, query.ObserverID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, domain.NewResourceNotFound("Article", "slug", query.Slug)
	}

	return article, nil
}

func (s *articleService) UpdateArticle(ctx context.Context, cmd port.UpdateArticleCommand) (*domain.Article, error) {
	// Get article and check if it exists
	article, err := s.articleRepo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, domain.NewResourceNotFound("Article", "slug", cmd.Slug)
	}

	// Check if user is author
	author, err := s.userRepo.FindByUsername(ctx, article.Author.Username)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if author == nil || author.ID != cmd.UserID {
		return nil, domain.NewForbiddenError("you are not the author of this article")
	}

	// Define duplicate check function
	checkDuplicate := func(title, slug string) error {
		// Check title
		existing, err := s.articleRepo.GetByTitle(ctx, title, nil)
		if err != nil {
			return domain.NewInternalError(err.Error())
		}
		if existing != nil && existing.ID != article.ID {
			return domain.NewAlreadyExistsError("Article with this title already exists")
		}

		// Check slug
		existing, err = s.articleRepo.GetBySlug(ctx, slug, nil)
		if err != nil {
			return domain.NewInternalError(err.Error())
		}
		if existing != nil && existing.ID != article.ID {
			return domain.NewAlreadyExistsError("Article with this slug already exists")
		}
		return nil
	}

	if err := article.Update(cmd.Title, cmd.Description, cmd.Body, checkDuplicate); err != nil {
		return nil, err
	}

	if err := s.articleRepo.Update(ctx, article); err != nil {
		return nil, domain.NewInternalError(err.Error())
	}

	return article, nil
}

func (s *articleService) DeleteArticle(ctx context.Context, slug string, userID int64) error {
	article, err := s.articleRepo.GetBySlug(ctx, slug, &userID)
	if err != nil {
		return domain.NewInternalError(err.Error())
	}
	if article == nil {
		return domain.NewResourceNotFound("Article", "slug", slug)
	}

	author, err := s.userRepo.FindByUsername(ctx, article.Author.Username)
	if err != nil {
		return domain.NewInternalError(err.Error())
	}
	if author == nil || author.ID != userID {
		return domain.NewForbiddenError("you are not the author of this article")
	}

	if err := s.articleRepo.Delete(ctx, article.ID); err != nil {
		return domain.NewInternalError(err.Error())
	}

	return nil
}

func (s *articleService) FavoriteArticle(ctx context.Context, slug string, userID int64) (*domain.Article, error) {
	article, err := s.articleRepo.GetBySlug(ctx, slug, &userID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, domain.NewResourceNotFound("Article", "slug", slug)
	}

	if err := s.articleRepo.Favorite(ctx, article.ID, userID); err != nil {
		return nil, domain.NewInternalError(err.Error())
	}

	return s.articleRepo.GetBySlug(ctx, slug, &userID)
}

func (s *articleService) UnfavoriteArticle(ctx context.Context, slug string, userID int64) (*domain.Article, error) {
	article, err := s.articleRepo.GetBySlug(ctx, slug, &userID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, domain.NewResourceNotFound("Article", "slug", slug)
	}

	if err := s.articleRepo.Unfavorite(ctx, article.ID, userID); err != nil {
		return nil, domain.NewInternalError(err.Error())
	}

	return s.articleRepo.GetBySlug(ctx, slug, &userID)
}

File: backend/typescript-node-express-hive/src/cells/article/article.service.ts
import { Article, slugify, type Tag } from "./article.domain.js";
import type {
    CreateArticleCommand,
    UpdateArticleCommand,
    DeleteArticleCommand,
    FavoriteArticleCommand,
    UnfavoriteArticleCommand,
    GetArticleQuery,
    ArticleService,
    ArticleRepository,
    UserProvider,
    GetArticlesQuery,
    ArticlesList,
    GetArticlesFeedQuery,
} from "./article.ports.js";
import {
    newAlreadyExistsError,
    newForbiddenError,
    newResourceNotFound,
} from "../../shared/errors/app-error.js";

export class DefaultArticleService implements ArticleService {
    constructor(
        private repo: ArticleRepository,
        private userProvider: UserProvider,
    ) {}

    async createArticle(cmd: CreateArticleCommand): Promise<Article> {
        const slug = slugify(cmd.title);
        const existing = await this.repo.getBySlug(slug);
        if (existing) {
            throw newAlreadyExistsError("Article with this title/slug already exists");
        }

        const author = await this.userProvider.getUser(cmd.authorId);
        const article = new Article(
            0,
            slug,
            cmd.title,
            cmd.description,
            cmd.body,
            cmd.tagList,
            new Date(),
            new Date(),
            false,
            0,
            {
                username: author.username,
                bio: author.bio,
                image: author.image,
                following: false,
            },
        );

        await this.repo.create(article, cmd.authorId);
        return article;
    }

    async getArticle(query: GetArticleQuery): Promise<Article> {
        const article = await this.repo.getBySlug(query.slug, query.observerId);
        if (!article) {
            throw newResourceNotFound("Article", "slug", query.slug);
        }
        return article;
    }

    async updateArticle(cmd: UpdateArticleCommand): Promise<Article> {
        const article = await this.getArticle({ slug: cmd.slug, observerId: cmd.userId });

        const author = await this.userProvider.getUserByUsername(article.author.username);
        if (author.id !== cmd.userId) {
            throw newForbiddenError("You are not the author of this article");
        }

        await article.update(
            {
                title: cmd.title,
                description: cmd.description,
                body: cmd.body,
            },
            async (_title, slug) => {
                const existing = await this.repo.getBySlug(slug);
                if (existing) {
                    throw newAlreadyExistsError("Article with this title/slug already exists");
                }
            },
        );

        await this.repo.update(article);
        return article;
    }

    async deleteArticle(cmd: DeleteArticleCommand): Promise<void> {
        const article = await this.getArticle({ slug: cmd.slug, observerId: cmd.userId });
        const author = await this.userProvider.getUserByUsername(article.author.username);
        if (author.id !== cmd.userId) {
            throw newForbiddenError("You are not the author of this article");
        }

        await this.repo.delete(article.id);
    }

    async favoriteArticle(cmd: FavoriteArticleCommand): Promise<Article> {
        const article = await this.getArticle({ slug: cmd.slug, observerId: cmd.userId });
        await this.repo.favorite(article.id, cmd.userId);
        return this.getArticle({ slug: cmd.slug, observerId: cmd.userId });
    }

    async unfavoriteArticle(cmd: UnfavoriteArticleCommand): Promise<Article> {
        const article = await this.getArticle({ slug: cmd.slug, observerId: cmd.userId });
        await this.repo.unfavorite(article.id, cmd.userId);
        return this.getArticle({ slug: cmd.slug, observerId: cmd.userId });
    }

    async getArticles(query: GetArticlesQuery): Promise<ArticlesList> {
        return this.repo.findAll(query);
    }

    async getFeed(query: GetArticlesFeedQuery): Promise<ArticlesList> {
        return this.repo.findFeed(query);
    }

    async getTags(): Promise<Tag[]> {
        return this.repo.findAllTags();
    }
}
